File: ios/Packages/FeatureServices/TransactionsService/GemstoneTransactionStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.AssetId
import typealias Gemstone.Transaction
import protocol Gemstone.GemTransactionStore
import GemstonePrimitives
import Preferences
import Primitives
import Store

public final class GemstoneTransactionStore: GemTransactionStore, @unchecked Sendable {
    private let store: TransactionStore

    public init(store: TransactionStore) {
        self.store = store
    }

    public func getSyncTimestamp(walletId: String, assetId: Gemstone.AssetId?) async throws -> UInt64 {
        let preferences = try WalletPreferences(walletId: WalletId.from(id: walletId))
        let timestamp = assetId.map { preferences.transactionsForAssetTimestamp(assetId: $0) } ?? preferences.transactionsTimestamp
        return UInt64(timestamp)
    }

    public func setSyncTimestamp(walletId: String, assetId: Gemstone.AssetId?, timestamp: UInt64) async throws {
        let preferences = try WalletPreferences(walletId: WalletId.from(id: walletId))
        if let assetId {
            preferences.setTransactionsForAssetTimestamp(assetId: assetId, value: Int(timestamp))
        } else {
            preferences.transactionsTimestamp = Int(timestamp)
        }
    }

    public func addTransactions(walletId: String, transactions: [Gemstone.Transaction]) async throws {
        try store.addTransactions(walletId: WalletId.from(id: walletId), transactions: transactions.map { try Primitives.Transaction($0) })
    }
}

File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/transactions/GemstoneTransactionStore.kt
package com.gemwallet.android.data.repositories.transactions

import com.gemwallet.android.cases.transactions.SaveTransactions
import com.gemwallet.android.data.service.store.WalletPreferencesFactory
import com.gemwallet.android.serializer.decodeJson
import com.wallet.core.primitives.Transaction
import com.wallet.core.primitives.WalletId
import uniffi.gemstone.GemTransactionStore

class GemstoneTransactionStore(
    private val walletPreferencesFactory: WalletPreferencesFactory,
    private val saveTransactions: SaveTransactions,
) : GemTransactionStore {

    override suspend fun getSyncTimestamp(walletId: String, assetId: String?): ULong {
        val preferences = walletPreferencesFactory.create(walletId)
        return (assetId?.let { preferences.transactionsForAssetTimestamp(it) } ?: preferences.transactionsTimestamp).toULong()
    }

    override suspend fun setSyncTimestamp(walletId: String, assetId: String?, timestamp: ULong) {
        val preferences = walletPreferencesFactory.create(walletId)
        if (assetId == null) {
            preferences.transactionsTimestamp = timestamp.toLong()
        } else {
            preferences.setTransactionsForAssetTimestamp(assetId, timestamp.toLong())
        }
    }

    override suspend fun saveTransactions(walletId: String, transactions: List<String>) =
        saveTransactions.saveTransactions(WalletId(walletId), transactions.map { it.decodeJson<Transaction>() })
}
